core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -662,6 +662,7 @@ private[spark] class TaskSetManager(
 
     val failureReason = s"Lost task ${info.id} in stage ${taskSet.id} (TID $tid, ${info.host}): " +
       reason.asInstanceOf[TaskFailedReason].toErrorString
+    var failureException: Option[Throwable] = None
     reason match {
```

Comment:
you could avoid using a `var` by making the exception the result of the entire match block, eg.

```scala
val failureException: Option[Throwable] = reason match {
 ...
```

honestly I don't have a strong opinion ... its nice to use a `val` so you know the value is set by the `match` block, but OTOH the `match` block is so big that is might be a bit hard to follow what is being returned ...